experimental/Schemes/SpaceGerms.jl

```diff
@@ -283,21 +981,275 @@ function Base.union(X::AbsSpaceGerm, Y::AbsSpaceGerm)
   Y = subscheme(underlying_scheme(X), I)
   return SpaceGerm(Y)
 end
 
+@doc raw"""
+    is_isolated_singularity(X::AbsSpaceGerm)
+
+Returns whether `(X,p)` has at most an isolated singularity.
+```jldoctest
+julia> X = affine_space(QQ,3);
+
+julia> R = coordinate_ring(X);
+
+julia> (x,y,z) = gens(R);
+
+julia> Q,_ = quo(R,ideal(R,[x^2+y^2]));
+
+julia> Y = SpaceGerm(Spec(Q),[0,0,0]);
+
+julia> is_isolated_singularity(Y)
+false
+
+```
+"""
+@attr Bool function is_isolated_singularity(X::AbsSpaceGerm)
+  dim(singular_locus(X)[1]) < 1 || return false
+  return true
+end
+
+##############################################################################
+# milnor_number, milnor_algebra for IHS
+# milnor_number for ICIS
+# -- beyond this we do no longer have a bouquet of spheres of same dimension
+##############################################################################
+@doc raw"""
+    milnor_algebra(X::HypersurfaceGerm)
+
+Returns the local Milnor algebra of `(X,p)` at p
+"""
+function milnor_algebra(X::HypersurfaceGerm)
+  R = localized_ring(OO(X))
+  ## milnor number independent of choice of representative
+  ## hence choose a polynomial representative for easier computation
+  f_poly = numerator(X.f)
+  I = ideal(R, R.([derivative(f_poly, i) for i=1:nvars(base_ring(R))]))
+  return quo(R,I)[1]
+end
+
+@doc raw"""
+    milnor_number(X::HypersurfaceGerm)
+    milnor_number(X::CompleteIntersectionGerm)
+
+Returns the local Milnor number of `(X,p)` at p
+"""
+function milnor_number(X::HypersurfaceGerm)
+  return vector_space_dimension(milnor_algebra(X))
+end
+
+function milnor_number(X::CompleteIntersectionGerm)
+  R = localized_ring(OO(X))
+  ## milnor number independent of choice of representative
+  ## hence choose polynomial representatives for easier computation
+  v = [numerator(a) for a in X.v]
+  w = typeof(v[1])[]   ## already used entries of v
+  dims = 0             ## for building up the alternating sum
+  sign = 1             ##     and the sign
+
+  ## alternating sum of the Le-Greuel formula, one summand per while-loop pass
+  while !is_empty(v)
+    found = false      ## keep track of 'colength condition satisfied?'
+
+    ## run through the potential choices, until colength condition satisfied 
+    for f in v
+      ## note: although we have already moved to polynomial data, we need to
+      ##       hand localized ring to helper to ensure shift to origin
+      ##       before computations
+      dtemp = _icis_milnor_helper(R,w,f)   ## colength; -1 if condition violated
+      if dtemp > -1
+        dims = dims + sign * dtemp         ## contribute to alternating sum
+        sign = -sign
+        push!(w,f)                          ## put f in 'used' list
+        deleteat!(v, findfirst(x->x==f,v)) ## remove f from 'unused' list
+        found = true
+        break
+      end
+    end
+    found == true || error("retry/general linear combinations not implemented yet")
```

Comment:
I think this should be doable. Can we assume `rand(kk, 0:10)`  to be implemented for our `base_ring` `kk`?

Reply:
Of course it is. 

I just did not want to make this PR even longer -- acutally I aimed to stay below 1000 lines for the reviewers' sake.